File: api/python/indigo/ml/hybridization.py
import math

OUTER_ELECTRONS = {
    "H": 1,
    "He": 2,
    "Li": 1,
    "Be": 2,
    "B": 3,
    "C": 4,
    "N": 5,
    "O": 6,
    "F": 7,
    "Ne": 8,
    "Na": 1,
    "Mg": 2,
    "Al": 3,
    "Si": 4,
    "P": 5,
    "S": 6,
    "Cl": 7,
    "Ar": 8,
    "K": 1,
    "Ca": 2,
    "Sc": 3,
    "Ti": 4,
    "V": 5,  # 4s2 3d3 (!)
    "Cr": 6,  # 4s1 3d5 (!)
    "Mn": 7,  # 4s2 3d5
    "Fe": 8,  # 4s2 3d6
    "Co": 9,
    "Ni": 10,
    "Cu": 1,
    "Zn": 2,
    "Ga": 3,
    "Ge": 4,
    "As": 5,
    "Se": 6,
    "Br": 7,
    "Kr": 8,
    "Rb": 1,
    "Sr": 2,
    "Y": 3,  # 5s2 4d1
    "Zr": 4,  # 5s2 4d2
    "Nb": 5,  # 5s1 4d4 (!)
    "Mo": 6,  # 5s1 4d5
    "Tc": 7,  # 5s2 4d5
    "Ru": 8,  # 5s1 4d7 (!!!)
    "Rh": 9,  # 5s1 4d8
    "Pd": 10,  # 4d10 (!)
    "Ag": 1,
    "Cd": 2,
    "In": 3,
    "Sn": 4,
    "Sb": 5,
    "Te": 6,
    "I": 7,
    "Xe": 8,
    "Cs": 1,
    "Ba": 2,
    "La": 3,  # 6s2 5d1
    "Ce": 4,  # 6s2 4f2 (!!!) Atomic num = 58, lantanoids
}

N_ORBS_HYBRIDIZATION = {
    1: "unhybridized",
    2: "sp",
    3: "sp2",
    4: "sp3",
    5: "sp3d",
    6: "sp3d2",
}


def num_bonds(atom):
    bonds = 0
    for nei in atom.iterateNeighbors():
        bonds += nei.bond().bondOrder()
    return bonds


def has_lone_pair(atom, n_bonds):
    return OUTER_ELECTRONS[atom.symbol()] - n_bonds >= 2


def lone_pairs(atom, n_bonds):
    outer_electrons = OUTER_ELECTRONS[atom.symbol()]
    pairs = math.floor((outer_electrons - n_bonds) / 2)
    return pairs


def cardon_hybridization(carbon):
    neighbors = carbon.degree()
    if neighbors == 4:
        return "sp3"
    if neighbors == 3:
        return "sp2"
    if neighbors <= 2:
        return "sp"
    return None


def match_minus_induction(atom):
    for nei in atom.iterateNeighbors():
        if nei.bond().bondOrder() in [2, 4]:
            return True
    return False


def oxygen_hybridization(oxygen):
    for nei in oxygen.iterateNeighbors():
        if nei.bond().bondOrder() == 2:
            return "sp2"
        if nei.bond().bondOrder() == 3:
            # for CO, but some sources do not recognize "sp" for oxygen
            return "sp"
        if nei.symbol() == "C" and match_minus_induction(nei):
            return "sp2"
    return "sp3"


def nitrogen_hybridization(nitrogen, n_bonds):
    n_orbs = nitrogen.degree() + has_lone_pair(nitrogen, n_bonds)
    minus_induction = False
    for nei in nitrogen.iterateNeighbors():
        if nei.symbol() == "C" and match_minus_induction(nei):
            minus_induction = True
    if n_orbs == 4 and minus_induction:
        return N_ORBS_HYBRIDIZATION[n_orbs - 1]
    return N_ORBS_HYBRIDIZATION[n_orbs]


def complex_hybridization(atom, neighbors):
    if neighbors == 4:
        if atom.symbol() in ["Pt", "Ni", "Cu", "Au", "Pd"]:
            return "sp2d"
        return "sp3"
    if neighbors == 5:
        return "sp3d"
    if neighbors == 6:
        return "sp3d2"
    if neighbors == 7:
        return "sp3d3"
    if neighbors == 8:
        return "sp3d4"
    return None


def get_hybridization(atom):
    # if the atomic number is undefined or ambiguous
    if atom.atomicNumber() == 0:
        return None

    # don't bother with the actinides and beyond
    elif atom.atomicNumber() >= 89:
        return "unhybridized"  # or "s"?
    if atom.atomicNumber() == 1:
        return "unhybridized"  # or "s"?

    n_bonds = num_bonds(atom)

    if n_bonds == 0:
        return "unhybridized"  # or "s"?

    if atom.symbol() == "C" and n_bonds >= 8:  # aromatic
        return "sp2"

    if atom.atomicNumber() == 6:
        return cardon_hybridization(atom)
    if atom.atomicNumber() == 8:
        return oxygen_hybridization(atom)
    if atom.atomicNumber() == 7:
        return nitrogen_hybridization(atom, n_bonds)

    if atom.degree() >= 4:
        return complex_hybridization(atom, atom.degree())

    # number of hybridized orbitals = number of neighbours + number of lone
    # electron pairs
    # H = N + L

    n_orbs = atom.degree() + lone_pairs(atom, n_bonds)
    return N_ORBS_HYBRIDIZATION[n_orbs]

File: core/indigo-core/molecule/src/hybridization.cpp
/****************************************************************************
 * Copyright (C) from 2009 to Present EPAM Systems.
 *
 * This file is part of Indigo toolkit.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/

#include "molecule/hybridization.h"

#include "molecule/elements.h"
#include "molecule/molecule.h"

namespace
{
    using namespace indigo;

    int getTotalBondsOrder(Molecule& molecule, int atomIndex)
    {
        const auto& atom = molecule.getVertex(atomIndex);
        int order = 0;
        for (auto neiIndex = atom.neiBegin(); neiIndex != atom.neiEnd(); neiIndex = atom.neiNext(neiIndex))
        {
            order += molecule.getBondOrder(atom.neiEdge(neiIndex));
        }
        order += molecule.getImplicitH(atomIndex);
        return order;
    }

    Hybridization getCarbonHybridization(Molecule& molecule, int atomIndex)
    {
        const auto numNeighbors = molecule.getVertex(atomIndex).degree() + molecule.getImplicitH(atomIndex);
        switch (numNeighbors)
        {
        case 4:
            return Hybridization::SP3;
        case 3:
            return Hybridization::SP2;
        case 2:
        case 1:
            return Hybridization::SP;
        default:
            throw HybridizationCalculator::Error("Couldn't calculate C hybridization properly");
        }
    }

    bool matchMinusInduction(Molecule& molecule, int atomIndex)
    {
        const auto& atom = molecule.getVertex(atomIndex);
        for (auto neiIndex = atom.neiBegin(); neiIndex != atom.neiEnd(); neiIndex = atom.neiNext(neiIndex))
        {
            const auto bondOrder = molecule.getBondOrder(atom.neiEdge(neiIndex));
            if (bondOrder == BOND_DOUBLE || bondOrder == BOND_AROMATIC)
            {
                return true;
            }
        }
        return false;
    }

    bool hasLonePair(Molecule& molecule, int atomIndex, int totalBondsOrder)
    {
        return Element::getNumOuterElectrons(molecule.getAtomNumber(atomIndex)) - totalBondsOrder >= 2;
    }

    int lonePairs(Molecule& molecule, int atomIndex, int totalBondsOrder)
    {
        const auto numOuterElectrons = Element::getNumOuterElectrons(molecule.getAtomNumber(atomIndex));
        return (numOuterElectrons - totalBondsOrder) / 2;
    }

    Hybridization getNitrogenHybridization(Molecule& molecule, int atomIndex, int totalBondsOrder)
    {
        const auto nOrbs = molecule.getVertex(atomIndex).degree() + static_cast<int>(hasLonePair(molecule, atomIndex, totalBondsOrder));
        bool minusInduction = false;
        const auto& atom = molecule.getVertex(atomIndex);
        for (auto neiIndex = atom.neiBegin(); neiIndex != atom.neiEnd(); neiIndex = atom.neiNext(neiIndex))
        {
            const auto neiVertexIndex = atom.neiVertex(neiIndex);
            if (molecule.getAtomNumber(neiVertexIndex) == ELEM_C && matchMinusInduction(molecule, neiVertexIndex))
            {
                minusInduction = true;
                break;
            }
        }
        if (nOrbs == 4 && minusInduction)
        {
            return Hybridization(nOrbs - 1);
        }
        if (nOrbs <= 4)
        {
            return Hybridization(nOrbs);
        }
        throw HybridizationCalculator::Error("Couldn't calculate N hybridization properly");
    }

    Hybridization getOxygenHybridization(Molecule& molecule, int atomIndex)
    {
        const auto& atom = molecule.getVertex(atomIndex);
        for (auto neiIndex = atom.neiBegin(); neiIndex < atom.neiEnd(); neiIndex = atom.neiNext(neiIndex))
        {
            const auto bondOrder = molecule.getBondOrder(atom.neiEdge(neiIndex));
            const auto neiVertexIndex = atom.neiVertex(neiIndex);

            if (bondOrder == BOND_DOUBLE)
            {
                return Hybridization::SP2;
            }
            if (bondOrder == BOND_TRIPLE)
            {
                // for CO, but some sources do not recognize "sp" for oxygen
                return Hybridization::SP;
            }

            if (molecule.getAtomNumber(neiVertexIndex) == ELEM_C && matchMinusInduction(molecule, neiVertexIndex))
            {
                return Hybridization::SP2;
            }
        }
        return Hybridization::SP3;
    }

    Hybridization getComplexHybridization(Molecule& molecule, int atomIndex)
    {
        const auto& numNeighbors = molecule.getVertex(atomIndex).degree();
        const auto atomNumber = molecule.getAtomNumber(atomIndex);
        switch (numNeighbors)
        {
        case 4:
            if (atomNumber == ELEM_Pt || atomNumber == ELEM_Ni || atomNumber == ELEM_Cu || atomNumber == ELEM_Au || atomNumber == ELEM_Pd)
            {
                return Hybridization::SP2D;
            }
            return Hybridization::SP3;
        case 5:
            return Hybridization::SP3D;
        case 6:
            return Hybridization::SP3D2;
        case 7:
            return Hybridization::SP3D3;
        case 8:
            return Hybridization::SP3D4;
        default:
            throw HybridizationCalculator::Error("Could not calculate %s hybridization properly", Element::toString(atomNumber));
        }
    }

    bool isAtomInAromaticRing(Molecule& rawMolecule, const int atomIndex)
    {
        Molecule molecule;
        molecule.clone(rawMolecule);
        if (!molecule.isAromatized())
        {
            molecule.aromatize({});
        }
        return molecule.getAtomAromaticity(atomIndex) == ATOM_AROMATIC;
        //        const auto& atom = molecule.getVertex(atomIndex);
        //        int order = 0;
        //        for (auto neiIndex = atom.neiBegin(); neiIndex != atom.neiEnd(); neiIndex = atom.neiNext(neiIndex))
        //        {
        //            if (molecule.getBondTopology(atom.neiEdge(neiIndex)) == TOPOLOGY_RING
        //        }
        return false;
    }
}

namespace indigo
{
    IMPL_ERROR(HybridizationCalculator, "HybridizationCalculator");

    Hybridization HybridizationCalculator::calculate(Molecule& molecule, int atomIndex)
    {
        const auto atomNumber = molecule.getAtomNumber(atomIndex);
        if (atomNumber == 0)
        {
            throw Error("Atomic number is undefined or ambiguous");
        }
        if (atomNumber >= 57)
        {
            throw Error("Hybridization calculation is not implemented for atomic numbers >= 57");
        }
        if (atomNumber == ELEM_H)
        {
            return Hybridization::S;
        }

        auto totalBondsOrder = getTotalBondsOrder(molecule, atomIndex);

        if (totalBondsOrder == 0)
        {
            return Hybridization::S;
        }

        if ((atomNumber == ELEM_C || atomNumber == ELEM_N || atomNumber == ELEM_O || atomNumber == ELEM_P || atomNumber == ELEM_S || atomNumber == ELEM_B) &&
            isAtomInAromaticRing(molecule, atomIndex))
        {
            return Hybridization::SP2;
        }

        if (atomNumber == ELEM_C)
        {
            return getCarbonHybridization(molecule, atomIndex);
        }
        if (atomNumber == ELEM_N)
        {
            return getNitrogenHybridization(molecule, atomIndex, totalBondsOrder);
        }
        if (atomNumber == ELEM_O)
        {
            return getOxygenHybridization(molecule, atomIndex);
        }

        if (molecule.getVertex(atomIndex).degree() >= 4)
        {
            return getComplexHybridization(molecule, atomIndex);
        }

        const auto nOrbs = molecule.getVertex(atomIndex).degree() + lonePairs(molecule, atomIndex, totalBondsOrder);
        return Hybridization(nOrbs);
    }
}
